java/code/src/com/redhat/rhn/frontend/action/LoginSetupAction.java

```diff
@@ -37,45 +41,55 @@
  */
 public class LoginSetupAction extends RhnAction {
 
-    private static Logger log = Logger.getLogger(LoginSetupAction.class);
+    private static final Logger LOG = Logger.getLogger(LoginSetupAction.class);
 
     /** {@inheritDoc} */
-    public ActionForward execute(ActionMapping mapping,
-        ActionForm form, HttpServletRequest request,
-        HttpServletResponse response) {
-
-        request.setAttribute("isUyuni", ConfigDefaults.get().isUyuni());
-        request.setAttribute("schemaUpgradeRequired",
-                LoginHelper.isSchemaUpgradeRequired().toString());
-        request.setAttribute("validationErrors", LoginHelper.validateDBVersion());
-
-        if (!UserManager.satelliteHasUsers()) {
-            return mapping.findForward("needuser");
+    public ActionForward execute(ActionMapping mapping, ActionForm form, HttpServletRequest request,
+                                 HttpServletResponse response) {
+        if (ConfigDefaults.get().isSingleSignOnEnabled() && SSOConfig.getSSOSettings().isPresent()) {
+            try {
+                Auth auth = new Auth(SSOConfig.getSSOSettings().get(), request, response);
+                auth.login("/rhn/YourRhn.do");
+            }
+            catch (SettingsException | IOException e) {
+               LOG.error(e.getMessage());
+            }
+            return null; // we forward the request to IdP via `auth.login`
         }
+        else {
+            request.setAttribute("isUyuni", ConfigDefaults.get().isUyuni());
+            request.setAttribute("schemaUpgradeRequired",
+                    LoginHelper.isSchemaUpgradeRequired().toString());
+            request.setAttribute("validationErrors", LoginHelper.validateDBVersion());
```

Comment:
This is in the `else` part (SSO is disabled), but I would hope that this path is still executed once we are successfully authenticated?

Reply:
This is the pre-login page before authentication: all the attributes here are used to show validation errors (for login) and schema upgrade messages.
Good catch: the schema upgrade message is only shown in the pre-login page, which means that we need to show that message somewhere if the user has enabled SSO.
Do you think we should display it after successful login in the landing page too?